src/xc_integrator/local_work_driver/device/cuda/kernels/uvvars.cu

```diff
@@ -209,4 +248,117 @@ void eval_uvvars_gga( size_t ntasks, size_t npts_total, int32_t nbf_max,
   );
 }
 
+
+
+
+
+
+
+
+
+
+template <int den_select>
+__global__ void eval_den_kern( size_t        ntasks,
+                                       XCDeviceTask* tasks_device ) {
+
+  const int batch_idx = blockIdx.z;
+  if( batch_idx >= ntasks ) return;
+
+  auto& task = tasks_device[ batch_idx ];
+
+  const auto npts            = task.npts;
+  const auto nbf             = task.bfn_screening.nbe;
+
+  double* den_eval_device   = nullptr;
+  // use the "U" variable (+/- for UKS) even though at this point the density (S/Z) is stored
+  switch( den_select ) {
+    case 0:
+      den_eval_device = task.den;
+      break;
+    case 1:
+      den_eval_device = task.den_pos;
+      break;
+    case 2:
+      den_eval_device = task.den_neg;
+      break;
+  }
+
+  const auto* basis_eval_device = task.bf;
+
+  const auto* den_basis_prod_device = task.zmat;
+
+  const int tid_x = blockIdx.x * blockDim.x + threadIdx.x;
+  const int tid_y = blockIdx.y * blockDim.y + threadIdx.y;
+
+  register double den_reg = 0.;
+
+  if( tid_x < nbf and tid_y < npts ) {
+
+    const double* bf_col   = basis_eval_device     + tid_x*npts;
+    const double* db_col   = den_basis_prod_device + tid_x*npts;
+
+    den_reg = bf_col[ tid_y ]   * db_col[ tid_y ];
+
+  }
+
+  // Warp blocks are stored col major
+  constexpr auto warp_size = cuda::warp_size;
+  //constexpr auto max_warps_per_thread_block = cuda::max_warps_per_thread_block;
+  den_reg = cuda::warp_reduce_sum<warp_size>( den_reg );
+
+
+  if( threadIdx.x == 0 and tid_y < npts ) {
+    atomicAdd( den_eval_device   + tid_y, den_reg );
+  }
+  
+
+}
+
+
+
+
+
+
+
+
+
+
+void eval_u_den( size_t ntasks, int32_t nbf_max, int32_t npts_max, density_id den_select,
+  XCDeviceTask* device_tasks, device_queue queue ) {
+
+  cudaStream_t stream = queue.queue_as<util::cuda_stream>();
+  dim3 threads( cuda::warp_size, cuda::max_warps_per_thread_block, 1 );
+  dim3 blocks( util::div_ceil( nbf_max,  threads.x ),
+               util::div_ceil( npts_max, threads.y ),
+               ntasks );
+
+  switch( den_select ){
+    case DEN:
+      eval_den_kern<0><<< blocks, threads, 0, stream >>>( ntasks, device_tasks );
+      break;
+    case DEN_S:
+      eval_den_kern<1><<< blocks, threads, 0, stream >>>( ntasks, device_tasks );
+      break;
+    case DEN_Z:
+      eval_den_kern<2><<< blocks, threads, 0, stream >>>( ntasks, device_tasks );
+      break;
+    default:
+      GAUXC_GENERIC_EXCEPTION("eval_u_den: den_select not properly set");
+  }
+
+}
+
+
+
+
+
+
+
+template __global__ void eval_den_kern<0>( size_t ntasks, XCDeviceTask* tasks );
```

Comment:
Was this required to compile? Shouldn't be... they should be instantiated via the above calls.